README.md
# MPXmda
Multivariable Discriminant Analysis open-source approach for the phase picker "Manneken Pix (MPX)".

**VERSION:** 2.5

**AUTHOR:**  Matteo Bagagli - <matteo.bagagli@erdw.ethz.ch> - [GitHub](https://github.com/billy4all)

**LICENSE:**  GNU-GPL v3

--------------------
## About
This collection of scripts follow the general rules descripted in Diehl2009,
in order to compute a Multivariate-Discriminant-Analysis (MDA) of the
Fisher coefficients for the 'MannekenPix Autopicker Algoritmh'.

This small routine has been carried out in order to give a possibility to users
to adopt an open-source approach instead of the macros and coomercial sofwares
suggested in the up-to-date guides [Alderson2004, Diehl2009, DiStefano2006].

This scripts should run smoothly in any Unix-like OS, as long as all
the necessary dependencies are correctly installed.
Starting from version 2.0, the scripts can be ran also in OSX.

## Dependencies
* Python v2.7.* or higher
* R
* Rscript

The needed package for R-statistics is called _DiscriMiner_.
To install it just type in an open R session:

```
> install.package(DiscriMiner)
```
## Getting Started
In order to execute MDA, the entire 'scripts' folder must be copied in your shell `PATH`. Then at shell promp just type:

```
$ MPXmda.sh -e MPXFILE -c CMNFILE [-wn]

  Where:
     -e ExportFile created by MPX
     -c CommandFile given in input to MPX
     -w TrueWeigth [Diehl2009] to be used in MDA (1,[2])
     -n Number of classes [4] % At the moment is tested only with 4 classes

  As output:
    - MannekenPix.cmn.OUT : an updated Command-file (*.cmn) version of ARG2,
                            containing the new Fisher Coefficients at the right
                            place. This file is ready-to-use for subsequent runs.
    - FisherCoeff.out : txt files with the results from lda in R
    - Format4MDA.csv : another formatting of *Exp for Excel/R analysis.

NB: The number of classes and the relative timing errors,
    MUST BE DEFINED INSIDE Format4MDA.py SCRIPT !!!
```

--------------------
### References
- Di Stefano, R et al. (2006). "Automatic seismic phase picking and consistent   observation error assessment: application to the Italian seismicity". In: Geophysical JournalInternational 165.1, pp. 121–134.
- Diehl, Tobias and Edi Kissling (2007). "Users guide for consistent phase picking at local to regional scales".
- Diehl, Tobias and Edi Kissling (2009). "Users Guide for MPX Picking System". In: PhD thesis–Appendix.
- Fisher, Ronald A (1936). "The use of multiple measurements in taxonomic problems". In: Annals of eugenics 7.2, pp. 179–18



CHANGELOG.md
# Changelog
---------------------------------
This GitHub repository has been initialized with the _v2.5_

* **v2.5** : Parsing input improved.
* **v2.0** : Added compatibility with OSX system (Mac)
* **v1.5** : Append to MPX_MDA which TrueWeight use [Diehl2009]
* **v1.0** : First release


# References 
---------------------------------
- Di Stefano, R et al. (2006). "Automatic seismic phase picking and consistent   observation error assessment: application to the Italian seismicity". In: Geophysical JournalInternational 165.1, pp. 121–134.
- Diehl, Tobias and Edi Kissling (2007). "Users guide for consistent phase picking at local to regional scales".
- Diehl, Tobias and Edi Kissling (2009). "Users Guide for MPX Picking System". In: PhD thesis–Appendix.
- Fisher, Ronald A (1936). "The use of multiple measurements in taxonomic problems". In: Annals of eugenics 7.2, pp. 179–18


scripts/Format4MDA.py
#!/usr/bin/env python
####
# This script format data in a proper way for a correct multivariate analysis
# of MannekenPix picker algorithm. Thi script format only PICKED phase by MPX.
# It returns the file Format4MDA.csv that can be loaded in R, SPSS or Excel
#
# USAGE:  Format4MDA.py *Exp
# AUTHOR: Matteo Bagagli @ ETH-Zurich // 012017

#    MPXmda : Multivariable Discriminant Analysis open-source approach
#             for the phase picker "Manneken Pix (MPX)"
#    Copyright (C) 2018  Matteo Bagagli
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.


# ******************************************************************************
# ********************* Reference Pick Class Uncertainties *********************
# ******************************************************************************
NumberOfClass=4
ClassBoundDict={
    '0':0.03,
    '1':0.06,
    '2':0.12,
    '3':0.24,
}
# ******************************************************************************


# --------------------------------------------------------------- Check & Vars
import sys,os,glob
if len(sys.argv)!=2:
    sys.stdout.write('USAGE:  Format4MDA.py *Exp'+os.linesep)
    sys.exit()
else:
    ExportFile=sys.argv[1]
# *Exp Field for extract values
AP_timeField=18  # start from 0 in *.Exp MPX file
AP_classField=19 # start from 0 in *.Exp MPX file
RP_timeField=25  # start from 0 in *.Exp MPX file
RP_classField=26 # start from 0 in *.Exp MPX file

# -------------------------------------------------------------------- Methods
def Pawk(stdIn,field,sep=" "):
    ''' Simil GNU-awk command for fields printing only
        Default field separator is space !!!
        IN: file-path or string.
        OUT: list of str object realted to the field specified.
        *** N.B: Enumeration for fields start from 0 , not 1 as awk does ...
    '''
    try:
        if os.path.isfile(stdIn):
            f=open(stdIn,"r")
            out=[]
            for line in f.readlines():
                fields=line.strip().split(sep)
                fields=filter(lambda x: x!='',fields)    # v1.0.3  weird loading issue if sep specified
                out.append(fields[field])
            f.close()
            return out
        else:
            fields=stdIn.strip().split(sep)
            fields=filter(lambda x: x!='',fields)       # v1.0.3  weird loading issue if sep specified
            return fields[field]
    except(IOError,NameError):
        sys.stderr.write('[Pawk] ERROR: Wrong input object ...'+os.linesep)
        return False
    except(IndexError):
        sys.stderr.write('[Pawk] ERROR: Field number out of range ...'+os.linesep)
        return False

def TrueWeigth_M1(value,classDict):
    ''' Methods that define the true weigth based on Absolute Time Errors
        between reference and MPX picks. (as described in DiStefano 2006)
    '''
    outVal=None
    if 0<=value and value<classDict['0']:
        outVal='0'
    elif classDict['0']<=value and value<classDict['1']:
        outVal='1'
    elif classDict['1']<=value and value<classDict['2']:
        outVal='2'
    elif classDict['2']<=value and value<classDict['3']:
        outVal='3'
    elif value>=classDict['3']:
        outVal='4'
    else:
        sys.stderr.write("TrueWeigth_M1: ERROR !!! No match ...")
        sys.stderr.write(os.linesep)
        sys.exit()
    return outVal

def TrueWeigth_M2(value,RefClass,classDict):
    ''' Methods that define the true weigth based on Absolute Time Errors
        between reference and MPX picks AND the reference Class. (as described in Diehl 2009).
        NB: *** RefClass is type:string !!!
    '''
    outVal=None
    if RefClass=='0':
        if 0<=value and value<classDict['0']:
            outVal='0'
        else:
            outVal='1'
    elif RefClass=='1':
        if classDict['0']<=value and value<classDict['1']:
            outVal='1'
        else:
            outVal='2'
    elif RefClass=='2':
        if classDict['1']<=value and value<classDict['2']:
            outVal='2'
        else:
            outVal='3'
    elif RefClass=='3':
        if classDict['2']<=value and value<classDict['3']:
            outVal='3'
        else:
            outVal='4'
    elif RefClass=='4': # Already rejected
        outVal='4'
    else:
        # If this error appear, check into yout *Exp file in search of
        # weird REFERENCE CLASS (field 27)
        sys.stderr.write(('[%s] ERROR !!! Unknown RefClass: %r'+os.linesep)%(
                            sys.argv[0].split('/')[-1],RefClass))
        sys.stderr.write(('[%s] ERROR !!! Unknown RefClass: %r'+os.linesep)%(
                            sys.argv[0].split('/')[-1],RefClass))
    return outVal

# ----------------------------------------------------------------------- Main
with open('Format4MDA.csv','w') as outID:
    # Writing the header for CSV  file.
    outID.write('FileID,MPX_tA,REF_tA,Ref_Class,AbsErr,WFilStoN,GpDlStoN,' \
                'GpDlAmpR,GpDlSigF,GpDlDelF,ThrCFRat,PcAboThr,PcBelThr,' \
                'CFNoiDev,TrueW_M1,TrueW_M2,TrueW_M3')
    outID.write(os.linesep)
    with open(ExportFile,'r') as expID:
        for ll,line in enumerate(expID):
            # Skip the first 14 lines of *Exp file
            if ll<=13:
                pass
            else:
                MPX_tA=Pawk(line,AP_timeField)
                # Switch if Picked by MPX, otherwise skip line
                if MPX_tA=='*******':
                    continue
                else:
                    FileID=Pawk(line,0)
                    REF_tA=Pawk(line,RP_timeField)
                    Ref_Class=Pawk(line,RP_classField)
                    AbsErr=abs(float(REF_tA)-float(MPX_tA))
                    WFilStoN=Pawk(line,28)
                    GpDlStoN=Pawk(line,29)
                    GpDlAmpR=Pawk(line,30)
                    GpDlSigF=Pawk(line,31)
                    GpDlDelF=Pawk(line,32)
                    ThrCFRat=Pawk(line,33)
                    PcAboThr=Pawk(line,34)
                    PcBelThr=Pawk(line,35)
                    CFNoiDev=Pawk(line,36)
                    TrueW_M1=TrueWeigth_M1(AbsErr,ClassBoundDict)
                    TrueW_M2=TrueWeigth_M2(AbsErr,Ref_Class,ClassBoundDict)
                    TrueW_M3='nan' # todo
                    outID.write('%s,%s,%s,%s,%f,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s' %(FileID,
                    MPX_tA,REF_tA,Ref_Class,AbsErr,WFilStoN,GpDlStoN,GpDlAmpR,GpDlSigF,GpDlDelF,
                    ThrCFRat,PcAboThr,PcBelThr,CFNoiDev,TrueW_M1,TrueW_M2,TrueW_M3))
                    outID.write(os.linesep)
#
sys.stdout.write('... DONE!'+os.linesep)

#################################################################
#       *** ADDITIONAL INFO ***
# When MannekenPix picks a seismogram, all predictors are usually populated even if
# some predictors are not used. Predictors are the variable values themselves while the
# coefficients of these variables are written in the command file. Unused predictors are
# expressed by coefficients of predictors equal to zero, not predictors necessarily equal to
# zero.


scripts/MPXmda.sh
#!/bin/bash
### MPX_mda.sh
# Simple routine that create a new command file for MPX updated with new Fisher
# coefficients. For more info read the README file. As a third parameter
# user must append TW1 or TW2 (TrueWeight 1 or 2) in order to apply the MDA on
# a particular class scheme definition. The user is referred @ Diehl,2009
#
# NB: The number of classes and the relative timing errors,
#     MUST BE DEFINED INSIDE Format4MDA.py SCRIPT !!!
#
# USAGE: MPX_mda.sh -e MPXFILE -c CMNFILE [-wn]
# AUTHOR: Matteo Bagagli @ ETH-Zurich , 01-2017
#

#    MPXmda : Multivariable Discriminant Analysis open-source approach
#             for the phase picker "Manneken Pix (MPX)"
#    Copyright (C) 2018  Matteo Bagagli
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.

VERSION="2.5"
# ----------------------------------------------------------------------- Check
if [ "$#" == "0" ] || [ "$1" == "-h" ]; then
  echo "[$(basename $0) v${VERSION}] USAGE: $(basename $0) -e MPXFILE -c CMNFILE [-wn]"
  echo "    Where:"
  echo "     -e ExportFile created by MPX"
  echo "     -c CommandFile given in input to MPX"
  echo "     -w TrueWeigth [Diehl2009] to be used in MDA (1,[2])"
  echo "     -n Number of classes [4]"
  echo "*** NB: The number of classes and the relative timing errors,"
  echo "        MUST BE DEFINED INSIDE Format4MDA.py SCRIPT !!!"
  exit
fi
# -------------------------------------------------------------------- Get Vars
while getopts 'e:c:w:n:' option
do
    case "${option}" in
            e) ExpFile="${OPTARG}";;
            c) CmnFile="${OPTARG}";;
            w) Weigth="${OPTARG}";;
            n) NClass="${OPTARG}";;
    esac
done
#
if [ -z "${ExpFile}" -o -z "${CmnFile}" ]; then
        echo "[$(basename $0) v${VERSION}] ERROR: need an *Exp AND a *cmn File !!!"
        exit
fi
# -------------------------------------------------------------------- Defaults
if [ -z "${Weigth}" ];then Weigth=2;fi
if [ -z "${NClass}" ];then NClass=4;fi
# ----------------------------------------------------------------- Expand Vars
#
if [ "${Weigth}" == "2" ]; then   # Preferred method
  weightfield=16
elif [ "${Weigth}" == "1" ]; then
  weightfield=15
else
  echo "[$(basename $0) v${VERSION}] ERROR: TrueWeight param. should be 1 or 2 !!!"
  exit
fi
# ------------------------------------------------------------------------ Main
echo "### Formatting"
Format4MDA.py ${ExpFile}
echo "### MDA"
### IN-File
echo "library(DiscriMiner)" > FisherMDA.R
echo "options(digits=5)" >> FisherMDA.R
echo "options(scipen=999)" >> FisherMDA.R        # To remove scientific notation
echo "rm(list=ls())" >> FisherMDA.R
echo "ExpFile <- read.csv('Format4MDA.csv',header=T)" >> FisherMDA.R
echo "colnames(ExpFile) <- c('FileID','MPX_tA','REF_tA','Ref_Class','AbsErr', \
                       'WFilStoN','GpDlStoN','GpDlAmpR','GpDlSigF','GpDlDelF', \
                       'ThrCFRat','PcAboThr','PcBelThr', 'CFNoiDev', \
                       'TrueW_M1','TrueW_M2','TrueW_M3')" >> FisherMDA.R
echo "df<-ExpFile[, c(${weightfield}, 6:14)]" >> FisherMDA.R
echo "df.lda2 <- linDA(df[,2:10], df[,1])" >> FisherMDA.R
echo "df.lda2\$functions" >> FisherMDA.R
# > summary(df.lda2) | If a summary of the analysis is wanted
###
Rscript FisherMDA.R > FisherCoeff.out

# ------------------------------------------------------------------ Out Format
sed '1d' FisherCoeff.out | awk '{sub($1 FS,"" );print}' > Fisher.cmn
# MAC+Linux v.2.0
cp ${CmnFile} MannekenPix.cmn.OUT
sed -i.trash "167s/.*/ $(sed  '1q;d' Fisher.cmn)/"  MannekenPix.cmn.OUT
sed -i.trash "170s/.*/ $(sed  '2q;d' Fisher.cmn)/"  MannekenPix.cmn.OUT
sed -i.trash "173s/.*/ $(sed  '3q;d' Fisher.cmn)/"  MannekenPix.cmn.OUT
sed -i.trash "176s/.*/ $(sed  '4q;d' Fisher.cmn)/"  MannekenPix.cmn.OUT
sed -i.trash "179s/.*/ $(sed  '5q;d' Fisher.cmn)/"  MannekenPix.cmn.OUT
sed -i.trash "182s/.*/ $(sed  '6q;d' Fisher.cmn)/"  MannekenPix.cmn.OUT
sed -i.trash "185s/.*/ $(sed  '7q;d' Fisher.cmn)/"  MannekenPix.cmn.OUT
sed -i.trash "188s/.*/ $(sed  '8q;d' Fisher.cmn)/"  MannekenPix.cmn.OUT
sed -i.trash "191s/.*/ $(sed  '9q;d' Fisher.cmn)/"  MannekenPix.cmn.OUT
sed -i.trash "194s/.*/ $(sed '10q;d' Fisher.cmn)/"  MannekenPix.cmn.OUT
#
rm Fisher.cmn FisherMDA.R MannekenPix.cmn.OUT.trash
#
echo "... DONE!"

# # ----::::----::::----::::----::::---- Improvement
# while getopts ':s:b:e:' option
# do
#     case "${option}"
#     in
#             s) Strain="${OPTARG}";;
#             b) Begin="${OPTARG}";;
#             e) End="${OPTARG}";;
#     esac
# done
# # ----::::----::::----::::----::::---- Check
# if [ -z "${Strain}" -o -z "${Begin}" -o -z "${End}" ]; then
#         echo "### MergeStrain2.0+: ERROR !!! All of the inputs required ..."
#         echo "###                  Type MergeStrain2.0+ without argouments "
#         echo "###                  for further help."
#         exit
# fi
